crates/sncast/src/helpers/data_transformer/transformer.rs

```diff
@@ -0,0 +1,161 @@
+        .get(function_selector)
+        .with_context(|| {
+            format!(
+                r#"Function with selector "{function_selector}" not found in ABI of the contract"#
+            )
+        })?;
+
+    let db = SimpleParserDatabase::default();
+
+    let result_for_cairo_expression_input =
+        process_as_cairo_expressions(calldata, function, &abi, &db)
+            .context("Error while processing Cairo-like calldata");
+
+    if result_for_cairo_expression_input.is_ok() {
+        return result_for_cairo_expression_input;
+    }
+
+    let result_for_serialized_input = process_as_serialized(calldata, &abi, &db)
+        .context("Error while processing serialized calldata");
+
+    match result_for_serialized_input {
+        Err(_) => result_for_cairo_expression_input,
+        ok => ok,
+    }
+}
+
+fn process_as_cairo_expressions(
+    calldata: &[String],
+    function: &AbiFunction,
+    abi: &[AbiEntry],
+    db: &SimpleParserDatabase,
+) -> Result<Vec<Felt>> {
+    let n_inputs = function.inputs.len();
+    let n_arguments = calldata.len();
+
+    ensure!(
+        n_inputs == n_arguments,
+        "Invalid number of arguments: passed {}, expected {}",
+        n_arguments,
+        n_inputs,
+    );
+
+    function
+        .inputs
+        .iter()
+        .zip(calldata)
+        .map(|(parameter, value)| {
+            let expr = parse(value, db)?;
+            let representation = build_representation(expr, &parameter.r#type, abi, db)?;
+            Ok(representation.serialize_to_vec())
+        })
+        .flatten_ok()
+        .collect::<Result<_>>()
+}
+
+fn process_as_serialized(
+    calldata: &[String],
+    abi: &[AbiEntry],
+    db: &SimpleParserDatabase,
+) -> Result<Vec<Felt>> {
+    calldata
+        .iter()
+        .map(|expression| {
+            let expr = parse(expression, db)?;
+            let representation = build_representation(expr, "felt252", abi, db)?;
+            Ok(representation.serialize_to_vec())
+        })
+        .flatten_ok()
+        .collect::<Result<_>>()
+}
+
+fn map_selectors_to_functions(abi: &[AbiEntry]) -> HashMap<Felt, AbiFunction> {
+    let mut map = HashMap::new();
+
+    for abi_entry in abi {
+        match abi_entry {
+            AbiEntry::Function(func) => {
+                map.insert(
+                    get_selector_from_name(func.name.as_str()).unwrap(),
+                    func.clone(),
+                );
+            }
+            AbiEntry::Constructor(constructor) => {
+                // Transparency of constructors and other functions
+                map.insert(
+                    get_selector_from_name(constructor.name.as_str()).unwrap(),
+                    AbiFunction {
+                        name: constructor.name.clone(),
+                        inputs: constructor.inputs.clone(),
+                        outputs: vec![],
+                        state_mutability: StateMutability::View,
+                    },
```

Comment:
Why do we have construct a separate abi entry for constructor?

Reply:
Constructor is deserialized as a dedicated `AbiEntry` variant. However, it's more handy for us to convert it to a regular function for selector transparency.

Reply:
Okay, maybe an explanation commend would be nice here

Reply:
I added an explaination